frontend/packages/console-shared/src/hooks/select-list.ts

```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { K8sResourceCommon } from '@console/internal/module/k8s';
+import { IRowData } from '@patternfly/react-table';
+
+export const useSelectList = <R extends K8sResourceCommon>(
+  visibleRows: R[],
+): {
+  onSelect: (
+    event: React.MouseEvent,
+    isSelected: boolean,
+    rowIndex: number,
+    rowData: IRowData,
+  ) => void;
+  selectedRows: R[];
+  setSelectedRows: React.Dispatch<React.SetStateAction<R[]>>;
+} => {
+  const [selectedRows, setSelectedRows] = React.useState<R[]>([]);
+
+  const onSelect = (
+    _event: React.MouseEvent,
+    isSelected: boolean,
+    rowIndex: number,
+    rowData: IRowData,
+  ) => {
+    const selectedUIDs = selectedRows?.map((node) => node.metadata.uid) ?? [];
+    const visibleUIDs = visibleRows?.map((row) => row.metadata.uid);
+    if (rowIndex === -1) {
+      if (isSelected) {
+        const uniqueUIDs = _.uniq([...visibleUIDs, ...selectedUIDs]);
+        setSelectedRows(
+          _.uniqBy(
+            [...visibleRows, ...selectedRows].filter((node) =>
+              uniqueUIDs.includes(node.metadata.uid),
+            ),
+            (n) => n.metadata.uid,
+          ),
+        );
+      } else {
+        setSelectedRows(
+          _.uniqBy(
+            selectedRows.filter((node) => !visibleUIDs.includes(node.metadata.uid)),
+            (n) => n.metadata.uid,
+          ),
+        );
```

Comment:
Maintain UIDs rather than K8sResource. If required create a Map to later map these UIDs to objects.